File: cp/bitwise_prod.cpp
#include <bits/stdc++.h>
#define num long long int 
using namespace std;


num find_z(num l,num r,num z)
{
    if(z>=l && z<=r)
    {
        return z;
    }
    else if(z<l)
    {
        z = (num)pow(2,(log(l)/log(2)));

        while(z<l)
            z+=(z/2);

        find_z(l,r,z);
    }
    else
    {

    }

}


int main() {
    // your code goes here
    num t;
    cin>>t;
    
    while(t--)
    {
        num x,y,l,r,z,m;

        cin>>x>>y>>l>>r;

        if(abs(l-r)<=100)
        {
            num maximum = (x & i)*(y & i); z=i;
            for(num i=l+1; i<=r; i++)
            {
                num tmp = (x & i)*(y & i);

                if(tmp > maximum)
                {
                    maximum = tmp; z=i;
                }
            }

            cout<<z<<"\n"; continue;
        }

        m = 

        z = (x | y);

        z = (num)pow(2,ceil(log(z+1)/log(2))) - 1;

        z = find_z(l,r,z);

        cout<<z<<"\n";
    }
    return 0;
}

File: cp/bitwise_prod.py
import math

t = int(input())

def check(l,r,x,y):

    maximum,z = (x & l)*(y & l) , l

    for i in range(l+1,r+1):

        tmp = (x & i)*(y & i)

        if(tmp > maximum):
            maximum , z = tmp , i

        if(maximum == (x*y)):

            return z

    return z

def find_z(l,r,x,y):

    z = x | y

    print(z)

    if(z>=l and z<=r):

        return z

    elif(z<l):

        while(z<l):

            z+= 2**(int(math.log2(z))+1)

        return z

    else:

        return check(l,r,x,y)

while t>0:

    t-=1

    x,y,l,r = input().split()
    x,y,l,r = int(x),int(y),int(l),int(r)

    z = x | y

    print(z, check(l,r,x,y))

    print(bin(x),bin(y),bin(z))
